src/webui/webapplication.cpp

```diff
@@ -173,45 +98,115 @@ namespace
+        bool isTranslationNeeded = !locale.startsWith("en") || locale.startsWith("en_AU") || locale.startsWith("en_GB");
+
+        while (i < data.size() && found) {
+            i = regex.indexIn(data, i);
+            if (i >= 0) {
+                //qDebug("Found translatable string: %s", regex.cap(1).toUtf8().data());
+                QByteArray word = regex.cap(1).toUtf8();
+
+                QString translation = word;
+                if (isTranslationNeeded) {
+                    QString context = regex.cap(4);
+                    translation = qApp->translate(context.toUtf8().constData(), word.constData(), 0, 1);
+                }
+                // Remove keyboard shortcuts
+                translation.replace(mnemonic, "");
+
+                // Use HTML code for quotes to prevent issues with JS
+                translation.replace("'", "&#39;");
+                translation.replace("\"", "&#34;");
+
+                data.replace(i, regex.matchedLength(), translation);
+                i += translation.length();
+            }
+            else {
+                found = false; // no more translatable strings
+            }
+        }
+    }
 }
 
-void WebApplication::action_public_index()
+WebApplication::WebApplication(QObject *parent)
+    : AbstractWebApplication(parent)
+    , m_isAltUIUsed(false)
 {
-    QString path;
-
-    if (!args_.isEmpty()) {
-        if (args_.back() == "favicon.ico")
-            path = ":/icons/skin/qbittorrent16.png";
-        else
-            path = WWW_FOLDER + args_.join("/");
+    for (int i = 0; i < metaObject()->methodCount(); ++i) {
+        QMetaMethod method = metaObject()->method(i);
+        if (QLatin1String(method.tag()) == QLatin1String("PUBLIC_API"))
+            m_publicAPIs << method.name();
     }
 
-    printFile(path);
+    configure();
+    connect(Preferences::instance(), &Preferences::changed, this, &WebApplication::configure);
 }
 
-void WebApplication::action_public_webui()
+void WebApplication::sendWebUIFile()
 {
-    if (!sessionActive())
-        printFile(PRIVATE_FOLDER + "login.html");
-    else
-        printFile(PRIVATE_FOLDER + "index.html");
+    if ((!m_isAltUIUsed) && (request().path.startsWith(PATH_PREFIX_IMAGES))) {
+        printFile(QLatin1String(":/icons/") + request().path.mid(PATH_PREFIX_IMAGES.size()));
+        header(Http::HEADER_CACHE_CONTROL, MAX_AGE_MONTH);
+        return;
+    }
+
+    const QString path {
+        (request().path != QLatin1String("/")
+                ? request().path
+                : (sessionActive()
+                   ? QLatin1String("/index.html")
+                   : QLatin1String("/login.html")))
+    };
+
+    // returns true if attempts to send the file is no longer needed
+    const auto sendFile {
+        [this](QString localPath) -> bool
+        {
+            QFileInfo fileInfo {localPath};
+            localPath = (fileInfo.isSymLink()
+                         ? fileInfo.symLinkTarget()
```

Comment:
What if this is symlink to a symlink?

Reply:
why not reject all symlink?

Reply:
Why not? (Since we went the way of restrictions.) I think, there shouldn't be serious reasons to use symbolic links here.